python3Packages.shapely: broken on darwin?

**Describe the bug**
python38Packages.shapely does not build for me on darwin. It looks like the previous release, 1.7.0, did build on hydra (https://hydra.nixos.org/build/133213521). I'm not sure if this is a difference depending on the version of shapely or on the version of macosx.

**To Reproduce**
`$ nix-shell -p python38Package.shapely`

**Failure**
```
$ nix-shell -p python38Package.shapely
[...]
running install tests
ImportError while loading conftest '/private/tmp/nix-build-python3.8-Shapely-1.7.1.drv-0/Shapely-1.7.1/tests/conftest.py'.
tests/__init__.py:4: in <module>
    from shapely.geos import geos_version_string, lgeos, WKTWriter
/nix/store/1l2vf26wvfsaddgm9w8ml8891bv0mv79-python3.8-Shapely-1.7.1/lib/python3.8/site-packages/shapely/geos.py:65: in <module>
    free = CDLL('/nix/store/alccnh6365ppv4690rinp7mswalm7gna-Libsystem-osx-10.12.6/lib/libc.dylib').free
/nix/store/c56kq0k3f3rdgkj3s1z5hw0j4ack9gdg-python3-3.8.6/lib/python3.8/ctypes/__init__.py:373: in __init__
    self._handle = _dlopen(self._name, mode)
E   OSError: dlopen(/nix/store/alccnh6365ppv4690rinp7mswalm7gna-Libsystem-osx-10.12.6/lib/libc.dylib, 6): image not found
builder for '/nix/store/vjidc9nms507pyvb7lnbm7gfhkv41m98-python3.8-Shapely-1.7.1.drv' failed with exit code 4
error: build of '/nix/store/vjidc9nms507pyvb7lnbm7gfhkv41m98-python3.8-Shapely-1.7.1.drv' failed
```

It looks like this store path is specifically being added by a nixpkgs patch, but that the patch refers to a file that doesn't exist.

**Additional context**
The're no libc.dylib here:
```
$ ls /nix/store/alccnh6365ppv4690rinp7mswalm7gna-Libsystem-osx-10.12.6/lib/
bundle1.o		crt1.o			lazydylib1.o		libdbm.tbd		libinfo.tbd		libproc.tbd		librpcsvc.tbd
crt0.o			dylib1.10.5.o		libSystem.B.tbd		libdl.tbd		libm.tbd		libpthread.tbd		libutil.tbd
crt1.10.5.o		dylib1.o		libSystem.tbd		libgcc_s.10.4.tbd	libmx.tbd		libresolv.9.dylib	system
crt1.10.6.o		gcrt1.o			libc.tbd		libgcc_s.10.5.tbd	libpoll.tbd		libresolv.dylib
```

**Notify maintainers**
@knedlsepp @dotlambda

**Metadata**
```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 16.7.0, macOS 10.12.6`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.10`
 - channels(mcgibbon): `"nixpkgs-21.03pre259239.39fe2fc4dc4"`
 - nixpkgs: `/Users/mcgibbon/.nix-defexpr/channels/nixpkgs`
 ```

--- a/pkgs/development/python-modules/shapely/default.nix
+++ b/pkgs/development/python-modules/shapely/default.nix
@@ -29,8 +29,7 @@ buildPythonPackage rec {
     (substituteAll {
       src = ./library-paths.patch;
       libgeos_c = GEOS_LIBRARY_PATH;
-      libc = "${stdenv.cc.libc}/lib/libc${stdenv.hostPlatform.extensions.sharedLibrary}"
-               + stdenv.lib.optionalString (!stdenv.isDarwin) ".6";
+      libc = stdenv.lib.optionalString (!stdenv.isDarwin) "${stdenv.cc.libc}/lib/libc${stdenv.hostPlatform.extensions.sharedLibrary}.6";
     })
   ];
 

--- a/pkgs/development/python-modules/shapely/library-paths.patch
+++ b/pkgs/development/python-modules/shapely/library-paths.patch
@@ -2,7 +2,7 @@ diff --git a/shapely/geos.py b/shapely/geos.py
 index d5a67d2..19b7ffc 100644
 --- a/shapely/geos.py
 +++ b/shapely/geos.py
-@@ -61,127 +61,10 @@ def load_dll(libname, fallbacks=None, mode=DEFAULT_MODE):
+@@ -61,127 +61,17 @@ def load_dll(libname, fallbacks=None, mode=DEFAULT_MODE):
              "Could not find lib {} or load any of its variants {}.".format(
                  libname, fallbacks or []))
  
@@ -128,7 +128,14 @@ index d5a67d2..19b7ffc 100644
 -    free.argtypes = [c_void_p]
 -    free.restype = None
 +_lgeos = CDLL('@libgeos_c@')
-+free = CDLL('@libc@').free
++if sys.platform == 'darwin':
++    # ctypes.CDLL(None) internally calls dlopen(NULL), and as the dlopen
++    # manpage says, "If filename is NULL, then the returned handle is for the
++    # main program". This way we can let the linker do the work to figure out
++    # which libc Python is actually using.
++    free = CDLL(None).free
++else:
++    free = CDLL('@libc@').free
 +free.argtypes = [c_void_p]
 +free.restype = None
  

